### ISSUE TITLE
Can't write non-overlapping blanket impls that involve associated type bindings

### ISSUE THREAD

**Issue by japaric** [rust-lang/rust#20400]
### STR

Example from libcore:

``` rust
#![crate_type = "lib"]
#![feature(associated_types)]
#![no_implicit_prelude]

trait Iterator {
    type Item;
}

trait AdditiveIterator {
    type Sum;

    fn sum(self) -> Self::Sum;
}

impl<I> AdditiveIterator for I where I: Iterator<Item=u8> {  //~error conflicting implementation
    type Sum = u8;

    fn sum(self) -> u8 { loop {} }
}

impl<I> AdditiveIterator for I where I: Iterator<Item=u16> {  //~note conflicting implementation here
    type Sum = u16;

    fn sum(self) -> u16 { loop {} }
}
```

### Version

7d4f487

No type can implement *both* `Iterator<Item=u8>` and `Iterator<Item=u16>`, therefore these blanket impls are non overlapping and should be accepted.

cc @nikomatsakis

**Comment by lfairy** [comment]
I encountered this bug too, while trying to optimize `ToString`:

```rust
pub enum Void {}

pub trait Display {
    // ...
    type AsStr = Void;
    pub fn fmt_as_str(&self) -> &<Self as Display>::AsStr { unimplemented!() }
}

pub trait ToString {
    fn to_string(&self) -> String;
}

impl<T: Display<AsStr=Void> + ?Sized> ToString for T {
    fn to_string(&self) -> String {
        format!("{}", self)
    }
}

impl<T: Display<AsStr=str> + ?Sized> ToString for T {
    fn to_string(&self) -> String {
        String::from_str(self.fmt_as_str())
    }
}
```

This change could speed up `"".to_string()` significantly (currently, `String::from_str` is between 2 and 5 times faster than `format!`) so I'd love to see this issue fixed.

**Comment by darinmorrison** [comment]
cc me

**Comment by edwardw** [comment]
cc me

**Comment by withoutboats** [comment]
cc

This is really causing problems for a crate I'm working on.

**Comment by cristicbz** [comment]
Is this bug likely to get any love? I keep hitting this (see #23341) too. I had a look around by I can't really follow how overlapping impls are determined, I would have liked to have a go at it.

**Comment by sgrif** [comment]
Any chance of this getting triaged any time soon? Also feeling pains from this issue.

**Comment by Kixunil** [comment]
I ran into this problem just now. I wanted to provide `and_then()` and `map()` methods for my type that would work if `Trait::Type` is `Option` or `Result`.

**Comment by ZerothLaw** [comment]
What's needed to advance this issue towards a fix? Are we dependent on Chalk?

**Comment by ZerothLaw** [comment]
Looks like you can kind of "trick" the compiler by using type specialization. On stable, 2015 edition: 

https://play.rust-lang.org/?gist=d803059c64390d7182445bb21756f98c&version=stable&mode=debug&edition=2015

**Comment by skade** [comment]
I'd like to add myself to the list of people that this causes problems for :).

**Comment by dbeckwith** [comment]
> Looks like you can kind of "trick" the compiler by using type specialization. On stable, 2015 edition:
> 
> https://play.rust-lang.org/?gist=d803059c64390d7182445bb21756f98c&version=stable&mode=debug&edition=2015

@ZerothLaw could you explain your trick a bit? I think I'm running into this issue and am hoping for a workaround but I'm struggling to understand your example. If you could give some advice on how to apply that to the code in #58171 that would be great.

**Comment by jonas-schievink** [comment]
To give an update on this: In 2016, [RFC 1672](https://github.com/rust-lang/rfcs/pull/1672) was proposed to fix this, but was postponed until the Chalk integration is done. Additionally, according to https://github.com/rust-lang/rfcs/pull/1672#issuecomment-262152934, allowing these kinds of impls would allow users to express mutually exclusive traits, which is a very powerful feature that needs more in-depth consideration by the language team (hence marking as blocked on an RFC as well).

(also adding the keyword "disjoint" to make this issue easier to find)

**Comment by steveklabnik** [comment]
Triage: these kinds of impls are still not accepted yet.

**Comment by SoniEx2** [comment]
We wanted  `FnMut(&char) -> bool` to be `Pattern` so you could use `char::is_ascii_digit` but this prevents that.

**Comment by hlbarber** [comment]
There is also a associated `const` version of this whereby

```
#![feature(associated_const_equality)]

trait Foo { const ENABLED: bool; }

trait Bar {}

struct S;

impl<T> Bar for T where T: Foo<ENABLED = true> {}
impl<T> Bar for T where T: Foo<ENABLED = false> {}
```

don't conflict.

**Comment by GoldsteinE** [comment]
Interestingly, it still fails with `-Z trait-solver=next`

**Comment by Ofenhed** [comment]
This is [possible on stable already](https://github.com/rust-lang/rfcs/pull/1672#issuecomment-1405377983), but not in a user friendly way.

**Comment by mversic** [comment]
[This crate](https://crates.io/crates/disjoint_impls) enables you to write disjoint implementations in a straightforward way

**Comment by rakshith-ravi** [comment]
@mversic  This looks great! Would it be possible for the same thing to work with functions as well?

Example:

```rust
pub trait RouterExt {
  fn mount_route<E: ApiEndpoint<Auth = NoAuth>>(self, endpoint: E) -> Self;
  fn mount_route<E: ApiEndpoint<Auth = SomeAuth>>(self, endpoint: E) -> Self;
}
```

This would allow different functions to be called depending on the associated type of the endpoint. Is this possible? I went through the tests, but couldn't find a similar example

**Comment by mversic** [comment]
> @mversic This looks great! Would it be possible for the same thing to work with functions as well?

no, you can't have two methods of the same name under one trait. I believe you should be able to do sth like this:

```rs
use disjoint_impls::disjoint_impls;

trait ApiEndpointDispatch {
    type Auth;
}

enum NoAuth {}
enum SomeAuth {}

struct Endpoint1 {}
impl ApiEndpointDispatch for Endpoint1 {
    type Auth = NoAuth;
}

struct Endpoint2 {}
impl ApiEndpointDispatch for Endpoint2 {
    type Auth = SomeAuth;
}

disjoint_impls! {
    pub trait ApiEndpoint {}
    impl<E: ApiEndpointDispatch<Auth = NoAuth>> ApiEndpoint for E {}
    impl<E: ApiEndpointDispatch<Auth = SomeAuth>> ApiEndpoint for E {}
}

pub trait RouterExt {
  fn mount_route<E: ApiEndpoint>(self, endpoint: E) -> Self;
}
```

**Comment by RodBurman** [comment]
This was raised in 2015 and here in 2024 I get:

cargo -V -v
cargo 1.81.0 (2dbb1af80 2024-08-20)
release: 1.81.0
commit-hash: 2dbb1af80a2914475ba76827a312e29cedfa6b2f
commit-date: 2024-08-20
host: aarch64-apple-darwin
libgit2: 1.8.1 (sys:0.19.0 vendored)
libcurl: 8.7.1 (sys:0.4.73+curl-8.8.0 system ssl:(SecureTransport) LibreSSL/3.3.6)
ssl: OpenSSL 1.1.1w  11 Sep 2023
os: Mac OS 15.0.1 [64-bit]
rod@Mac nonover % cargo build
   Compiling nonover v0.1.0 (/Users/rod/code/rust/triage/nonover)
error[E0119]: conflicting implementations of trait `AdditiveIterator`
  --> src/lib.rs:21:1
   |
15 | impl<I> AdditiveIterator for I where I: Iterator<Item=u8> {  //~error conflicting implementation
   | --------------------------------------------------------- first implementation here
...
21 | impl<I> AdditiveIterator for I where I: Iterator<Item=u16> {  //~note conflicting implementation here
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation
For more information about this error, try `rustc --explain E0119`.
error: could not compile `nonover` (lib) due to 1 previous error

So the "bug" or lack of feature remains.

**Comment by QuineDot** [comment]
For those expecting it to work with the new trait solver automatically, [here's a reminder that it's blocked on an RFC](https://github.com/rust-lang/rust/issues/20400#issuecomment-493206721) as it unlocks mutually exclusive traits.
